Data protection for sensitive files

Venemo opened the pull request:
Do not merge yet, this is incomplete. Only submitting to get early feedback for the code.

Venemo commented:
What remains to be done:

* I'll add a bunch of tests to make sure the implementation works. This will also undoubtedly reveal a bunch of bugs in there.
* Add support for processing the legacy way of storing the data.  Simple enough.
* Then, hook up the master lock code and the salt data to use the new class.

Will continue working on this tonight and hopefully will finish tomorrow.

chriadam commented:
The approach seems good to me, but I don't know much about this stuff, so if other people could comment that would be appreciated.  In any case, this form of protection is definitely better than what we have now for those two files.

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,270 @@
+
+    // Open all files
+    bool allFilesLost = std::all_of(files.begin(), files.end(), [](QSharedPointer<QFile> &file) {
+        bool cantopen = file->open(QIODevice::ReadOnly);
```
bool opened variable is now somehow upside down.

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,270 @@
+        QDir oldDataDir(entry.absolutePath());
+        oldDataDir.removeRecursively();
+    }
+
```
Empty line

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,270 @@
+        qCDebug(lcSailfishSecretsDaemon) << "Protected root directory didn't exist, so the data is empty.";
+    }
+
+    auto oldDirectoryEntries = dir.entryInfoList(QDir::Dirs);
```
autos are not helping here more like hindering legibility

monich replied:
That's their purpose isn't it?

Venemo replied:
Sorry.  
I know you guys hate `auto` and will remove them when I finalize this PR.

For me personally, it actually improves legibility.

monich replied:
Your own code is always readable for yourself, at least right after you wrote it. There's nothing surprising about that ))

Venemo commented:
Pushed what I have now. Almost there!

Done:

* Fixed review findings
* Added unit tests
* Fixed issues found by unit tests
* Hooked up to salt data

Not yet:

* Need to run all tests on device to make sure the daemon behaves sanely
* Need to hook up the masterlock as well

Venemo commented:
Ok, now refactored a bit and hooked up the lock code data as well. Now ready for review.

Venemo commented:
I don't have my dev device at hand right now, but will run all tests on it once I get home before merging all this.

Venemo commented:
@dcaliste I couldn't add you as a reviewer but of course your opinion is welcome here too.

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,326 @@
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory:" << m_path;
```
Use different wording here so that getData and putData are not logging the same error message e.g. "Can't create protected root directory when writing:" << m_path;

Venemo replied:
Will fix.

rainemak commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -747,12 +746,20 @@ QByteArray Daemon::ApiImpl::SecretsRequestQueue::saltData() const
+    QByteArray saltData;
+    DataProtector::Status s = dataProtector.getData(&saltData);
+    if (s != DataProtector::Success) {
+        qWarning(lcSailfishSecretsDaemon) << "Salt data is corrupted.";
```
Again wording make this distinguishable from line 778.

rainemak replied:
When initialsalt is a file e.i. one is upgrading secrets => it's always "Salt data is corrupted". Making in practice previous secrets void. Maybe we can live with this as we haven't never related this before.

Venemo replied:
Well, do you want to support this or not? If yes, then there is a commit that deals with it. If not, then just accept that you have to delete the old data when you upgrade.

rainemak commented on tests/Secrets/tst_dataprotection/tst_dataprotection.h:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2017 Jolla Ltd.
+ * Contact: Chris Adams <chris.adams@jollamobile.com>
```
Wrong contact and year

Venemo replied:
We use the same header everywhere. Did not want to alter it just for this one file.

rainemak replied:
But this still not Chris' rather yours.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,326 @@
+            qCWarning(lcSailfishSecretsDaemon) << "Can't write file:" << file->fileName();
+            return ErrorCannotWriteFile;
+        }
+        file->commit();
```
Not a big issue, but you're not checking that the commit operation succeed.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,326 @@
+        return ErrorInvalidPath;
+    }
+
+    QString parentPath = m_path.mid(0, slashIndex);
```
Shouldn't you use a ```QFileInfo(m_path).path()``` to do this? I understand that there is no chance to have '/' character in the name of the legacy file, but in case, and it's more syntaxically correct, I guess.

Venemo replied:
Yep, good idea.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,326 @@
+    QByteArray fileContents = legacyFile.readAll();
+
+    // Store the data in a redundant manner
+    putData(fileContents);
```
putData() may return an error here, no? Then the legacy file should not be removed.

dcaliste commented on daemon/SecretsImpl/dataprotector_p.h:
```diff
@@ -0,0 +1,67 @@
+    };
+    Q_ENUM(Status)
+
+    explicit DataProtector(const QString &path, QObject *parent = Q_NULLPTR);
```
Just a question, or suggestion, would it be clearer that path should be a directory if the constructor argument would be a QDir?

rainemak commented:
Please rebase

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,326 @@
+}
+
+
+DataProtector::Status DataProtector::processLegacy()
```
This feels a bit wrong to me as we have not released this yet. Hence, I'd just have a disclaimer in commit message that  old legacy files must be nuked e.i. rm -rf privileged/Secrets

Venemo replied:
Okay, I can get rid of this.

Venemo replied:
I think the fact that it's a directory is just an implementation detail. Thus I would advise not to have the calling code depend on that detail. This data protection solution is only what we could come up with in such a short notice, but we will refine it when/if we have a better idea and time to implement it.

Venemo commented:
Rebased on top of latest master. Fixed review findings.  
At the end we agreed with @rainemak to just get rid of the "legacy" data handling, as it makes very little sense to support something that was never actually released.

I'll run tests on my device and see how it goes from there.

rainemak commented on daemon/SecretsImpl/secrets.cpp:
```diff
@@ -681,26 +682,31 @@ bool Daemon::ApiImpl::SecretsRequestQueue::compareTestCipherText(
+    if (s != DataProtector::Success) {
+        qCWarning(lcSailfishSecretsDaemon) << "Can't read lock code data, assuming it's corrupted. DataProtector returned:" << s;
+        qCWarning(lcSailfishSecretsDaemon) << "NOTE: If you are running a pre-release sailfish-secrets version, you need to delete all old data before proceeding";
+        abort();
```
I'm not sure whether aborting secrets is good upon failure without letting user to know. How do you recover from the error?

rainemak replied:
Let's go with this for now and improve later. Fyi @chriadam

Venemo replied:
My suggestion:

* For 2.2.0 let's just go with `abort()`. Alternatives: automatically delete the data when it's corrupted, or just let the daemon run even though anything that used the lock code (or salt) data won't work anymore.
* For 2.2.1 and onwards: we should add a user-friendly dialog, maybe a system prompt, that lets the user know what happened and offer to delete all the corrupted data.

Venemo replied:
For now I will add a `TODO` comment as a reminder to not forget this later.

rainemak commented on tests/Secrets/tst_dataprotection/tst_dataprotection.cpp:
```diff
@@ -0,0 +1,119 @@
+
```
Extra empty line

Venemo replied:
Fixed.

rainemak commented on daemon/SecretsImpl/dataprotector_p.h:
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2018 Jolla Ltd.
+ * Contact: Chris Adams <chris.adams@jollamobile.com>
```
Same as tst_ file

rainemak commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,265 @@
+
```
Missing copyright header

Venemo replied:
Added now.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,271 @@
+    bool allOpen = std::all_of(files.begin(), files.end(), [](QSharedPointer<QSaveFile> &file) {
+        bool isOpen = file->open(QIODevice::WriteOnly);
+        if (!isOpen) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName();
```
Sorry to add comments late, if you still have time, it would be nice to log the error string, so it's easier to debug or understand the reason of the error:
```qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName() << file->errorString();```

Venemo replied:
Sorry, you were literally 5 minutes late with this. But that's a nice catch and I'll add it next chance I get.

dcaliste replied:
Yep, thought about it at lunch time, but was busy to report it before… Anyway, it's not critic, and can definitively come with a later commit.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,271 @@
+    for (QSharedPointer<QSaveFile> &file : files) {
+        qint64 bytesWritten = file->write(bytes);
+        if (bytesWritten != bytes.size()) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't write file:" << file->fileName();
```
Same here.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,271 @@
+            return ErrorCannotWriteFile;
+        }
+        if (!file->commit()) {
+            qCWarning(lcSailfishSecretsDaemon) << "Could not commit file:" << file->fileName();
```
And also here.

dcaliste commented on daemon/SecretsImpl/dataprotector.cpp:
```diff
@@ -0,0 +1,271 @@
+        bool isOpen = file->isOpen();
+        if (!isOpen) {
+            // Loss of any individual file here doesn't yet necessarily mean an error
+            qWarning(lcSailfishSecretsDaemon) << "can't open file:" << file->fileName();
```
Same here to use ```errorString()```, see the comments in putData() part.
